builtin/providers/cloudstack/resource_cloudstack_template.go

```diff
@@ -219,9 +236,16 @@ func resourceCloudStackTemplateRead(d *schema.ResourceData, meta interface{}) er
 	d.Set("password_enabled", t.Passwordenabled)
 	d.Set("is_ready", t.Isready)
 
-	setValueOrUUID(d, "os_type", t.Ostypename, t.Ostypeid)
-	setValueOrUUID(d, "zone", t.Zonename, t.Zoneid)
+	setValueOrUUID(d, "project", t.Project, t.Projectid)
 
+	if t.Zoneid == UnlimitedResourceID {
+		setValueOrUUID(d, "zone", t.Zonename, UnlimitedResourceID)
+	} else {
+		setValueOrUUID(d, "zone", t.Zonename, t.Zoneid)
+	}
```

Comment:
Ah, yes, I see the error.

The main issue here is CloudStack does not return a `-1` when you read from the API. Setting it to `-1` results in an empty response for `t.Zoneid` when reading, so Terraform considers it a change and tries to re-provision every time. The if statement should read:

```
	if t.Zoneid == "" {
		setValueOrUUID(d, "zone", t.Zonename, UnlimitedResourceID)
	} else {
		setValueOrUUID(d, "zone", t.Zonename, t.Zoneid)
	}
```

Does that make more sense?